### DISCUSSION TITLE
Lora module missing packets in channel scan mode...but no missing packet in contineous receive mode

### DISCUSSION THREAD

**Discussion by codev123** [jgromes/RadioLib#940] [Q&A]
thank sir for amazing work! i have been recently implementing a repeater node that receives lora packets from multiple Lora transmitters, and retransmitting when channel is free.
The issue i am encountering at this repeater code is that, i am missing packets using this scan method, while i dont miss any packet when i receive packets from multiple transmitters on a receiver that is always in interrupt based receive mode. can you please help me what i am doing wrong.

here is various flags to manage tansmission, reception and scan flags with loop code........
Repeater code:

```c++
// save transmission states between loops
int transmissionState = RADIOLIB_ERR_NONE;
// flag to indicate transmission state
bool transmitFlag = false;
// flag to indicate reception state
bool receiveFlag = false;
// flag to indicate that a packet was received or transmitted or channel scan
volatile bool scanFlag = false;
void setFlag(void) {
  // we sent or received a packet or scanned channel, set the flag
  scanFlag = true;
}

void loop() {
  // check if the flag is set
  if (scanFlag) {
    int state = RADIOLIB_ERR_NONE;
    scanFlag = false;  // reset flag
    //if the previous operation was transmission, check for response and print the result
    if (transmitFlag) {
      if (transmissionState == RADIOLIB_ERR_NONE) {
        digitalWrite(txStatus, HIGH);  //LED OFF when packet transmitted
        Serial.println(F("transmission finished!"));
      } else {
        Serial.print(F("failed, code "));
        Serial.println(transmissionState);
      }
      transmitFlag = false;  //reset the Transmission flag
    } else if (receiveFlag) {
      lastDataTime = millis();  // If data is received, update the timestamp
      //read received packet
      int numBytes = sizeof(DataPacket);  // Set the size to the structure size
      state = radio.readData((uint8_t*)&ReceivedPacket, numBytes);
      if ((state == RADIOLIB_ERR_NONE) && (ReceivedPacket.repeaterID == NodeID ) {
          digitalWrite(rxStatus, LOW);  //LED ON when packet received
          //delay(100);
          //Process received data packet details
          Serial.println(F("[SX1268] Received Data packet details................................."));
          Serial.print(F("sensor ID: "));
          Serial.println(ReceivedPacket.sensorID);
          Serial.print(F("pressure: "));
          Serial.println(ReceivedPacket.pressure);
          
          digitalWrite(rxStatus, HIGH);  //LED OFF when packet received
          //assign received packet to transmit packet buffer to retransmit
          ReceivedPacket.repeaterID = RepeaterID;
          if (bufferIndex < bufferSize) {
            TransmitPacket[bufferIndex] = ReceivedPacket;
            bufferIndex++;
            Serial.print(F("Buffer index: "));
            Serial.println(bufferIndex);
          } else {
            ;// Buffer is full, handle accordingly (e.g., drop the packet or wait)
          }
        } else if (state == RADIOLIB_ERR_RX_TIMEOUT) {
        Serial.println(F("Timed out while waiting for packet!"));
        digitalWrite(rxStatus, HIGH);  //LED OFF when packet received
      } else {
        Serial.println(F("Failed to receive packet, code "));
        Serial.println(state);
      }
      receiveFlag = false;
    } else {
      // check CAD result
      state = radio.getChannelScanResult();
      if (state == RADIOLIB_LORA_DETECTED) {
        // LoRa packet was detected
        Serial.print(F("Packet detected, starting reception... "));
        state = radio.startReceive();
        if (state == RADIOLIB_ERR_NONE) {
          Serial.println(F("success!"));
        } else {
          Serial.print(F("reception failed, code "));
          Serial.println(state);
        }
        // set the flag for ongoing reception
        receiveFlag = true;
      } else if (state == RADIOLIB_CHANNEL_FREE) {  //channel free
        //Serial.println(F("Channel is free!"));
        //received packet, send it to next node
        if (bufferIndex > 0) {
          Serial.print(F("Transmission started..... "));
          digitalWrite(txStatus, LOW);  //LED ON when packet transmission started
          transmissionState = radio.startTransmit((uint8_t*)&TransmitPacket[0], sizeof(DataPacket));
          transmitFlag = true;
          // Shift the remaining elements in the buffer
          for (int i = 0; i < bufferIndex - 1; i++) {
            TransmitPacket[i] = TransmitPacket[i + 1];
          }
          bufferIndex--;
        }
      } else {
        //Serial.print(F("Channel scan Failed, code "));
        //Serial.println(state);
      }
    }
    // if we're not receiving, start scanning again
    if (!receiveFlag && !transmitFlag) {
      //Serial.print(F("Not in Rx/Tx mode....starting for Lora preamble...."));
      state = radio.startChannelScan();
      if (state == RADIOLIB_ERR_NONE) {
        //Serial.println(F("success!"));
      } else {
        //Serial.print(F("failed, code "));
        //Serial.println(state);
      }
    }
  }
}
```
the receiver code :

```c++
// flag to indicate that a packet was received
volatile bool receivedFlag = false;
void setFlag(void) {
  // we got a packet, set the flag
  receivedFlag = true;
}

void loop() {
  // check if the flag is set
  if (receivedFlag) {
    // If data is received, update the timestamp
    lastDataTime = millis();
    // reset flag
    receivedFlag = false;
    //read received packet
    int numBytes = sizeof(DataPacket);  // Set the size to the structure size
    int state = radio.readData((uint8_t*)&ReceivedPacket, numBytes);
    if ((state == RADIOLIB_ERR_NONE) && ReceivedPacket.repeaterID == RepeaterID) {
      //Process received data packet details
      Serial.println(F("[SX1268] Received Data packet details.............................................."));
      Serial.print(F("[SX1268] RSSI:\t\t"));
      Serial.print(radio.getRSSI());
      Serial.println(F(" dBm"));
      Serial.print(F("[SX1268] SNR:\t\t"));
      Serial.print(radio.getSNR());
      Serial.println(F(" dB"));
    } else {
      Serial.println(F("Received packet is not for receiver"));
      digitalWrite(rxStatus, HIGH);  //LED OFF when packet not received
    }
}
```

so i receive all transmitter packets(Lora transmitters are transmitting when channel is free) using this receiver code but using repeater code i miss packets when time between packets is small.

also, when this issue happen, the preamble is detected, but packet is received after 5-10 seconds later

```
16:22:40.915 -> Packet detected, starting reception... success!
16:22:57.875 -> [SX1268] Received Data packet details.................................
16:22:57.875 -> sensor ID: 1
16:22:57.875 -> pressure: 214
```

here packet is detected at 16:22:40, but scanFlag interrupt was active after 17 seconds to receive packet at 16:22:57, any packet received in these 17 seconds will be lost.

normally it should be like this 
```
16:22:40.145 -> Packet detected, starting reception... success!
16:22:40.440 -> [SX1268] Received Data packet details.................................
16:22:40.440 -> sensor ID: 2
16:22:40.440 -> pressure: 208
```

**Comment by jgromes** [comment]
Since the repeater code is quite a lot more complex, have you measured how long it takes to switch between packet detection and reception? I don't see any obvious reason why channel scan should miss packets that are received by the receiver code. However, you do have a lot of printing to the Serial port, and this: `TransmitPacket[i] = TransmitPacket[i + 1];` - that is copying all of your packets, instead of e.g. shifting a pointer. No idea how large `DataPacket` is though.

> also, when this issue happen, the preamble is detected, but packet is received after 5-10 seconds later

I have to guess here, but I'd say that's the next packet. I think that you are not fast enough when switching from CAD to receive. But overall this does't seem like an issue in the library itself, so I will convert this to a discussion instead.

**Reply by codev123** [reply]
thank you sir for quick reply... the time between packet detection and packet reception is actually puzzling me. sometimes its very short as i have shown serial print text...from milliseconds to 30 seconds. while i observed at receiver code that in between these 30 seconds, receiver has actually received 2 packets.
also, i have tried with disabled seriable printing... still same issue

shifting packets ....  TransmitPacket[i] = TransmitPacket[i + 1];...... this step i am doing after transmission, packet size is 10 bytes only.... and its not taking any time.........

as you can see i am directly jumping from packet detection i.e. CAD to packet reception, no operation in between still i am encountering this issue.

can you please guide me how to resolve this issue. i dont think its libraray problem, it could be some code handling issue.

Could it be due to multipaths or signal conditioning, as lora detects packets by scanning for preamble, could it be that it doesnt receive any packet due to packet loss, and then when it receives next packet, it process it, but how can receiver receive those 2 packets without packet loss?

tommorow i will upload this issue using both receiver and repeater codes Serial prints, that will help you visualise the issue easily. thank you sir

**Reply by StevenCellist** [reply]
Definitely not an RF expert, but I did try my hand on channel scan (CS) activity for a few days when working on the LoRaWAN stack. My personal experience was that using CS was fine for higher SF (10+), but using lower SF it became unusable. Statistics on the web show ~90% accuracy in channel activity detection, but on low SF I almost never caught a package - did get detections, apparently couldn't catch the actual payload in time.

**Reply by codev123** [reply]
Exactly the issue. It sometimes detect a packet and receive payload in no time i.e. milliseconds while other times it takes quite longer times as dio1 pin don't become active to receive payload as it's in startreceive mode. It only becomes active when next packet is received as radio is in start receive mode.

While I have tried setting setdio1actionmode(setflag) for simply receiving and transmitting, and  there were not any issue. Only when added scan method on same flag.

I have also tried using setpacketreceiveaction(setflag), i didn't encounter any issue, issue only occurs at channel scan, it detects a packet but doesn't always receive the packet/activate the flag to receive the payload.

Can it work on spreading factor 10, or i have to work with sf of 11 and 12, as I don't want to increase transmission time and decrease range by increasing bandwidth.

Also is there any function to calculate transmission time in library

Update: funny issue, I don't have any Lora transmitter at 434.0Mhz frequency and when i started channel scan at repeater it detects packet but never receive as there is no packet. At spreading factor <=10 this happening very much, but at sf >10 radio still detects packet but at less frequency. 

It seems like a chip specific issue. @jgromes sir, is there any other way to avoid collision of radio packets
Thankyou sir

**Reply by GUVWAF** [reply]
I’ve had the same issue of missing packets when using `startReceive()` just after `scanChannel()`. The solution I ended up using is to set additional IRQ flags like this: `startReceive(RADIOLIB_SX126X_RX_TIMEOUT_INF, RADIOLIB_SX126X_IRQ_RX_DEFAULT | RADIOLIB_SX126X_IRQ_PREAMBLE_DETECTED |
RADIOLIB_SX126X_IRQ_HEADER_VALID)`. 
Before transmitting, you then first check with `getIrqStatus()` whether either `PREAMBLE_DETECTED` or `HEADER_VALID` is set. One caveat with this: it seems to capture quite some false preamble detections, so I ended up using a timer to clear the `PREAMBLE_DETECTED` flag when it should have received a header given the preamble duration. Header detection seems very reliable in my testing.

_Edit_: I later found that this might be because the CAD exit mode is set to “go to standby” instead of “remain in Rx”: 
https://github.com/jgromes/RadioLib/blob/86136d3b46469f4a16d804c052bc2ca8ac3d62d6/src/modules/SX126x/SX126x.cpp#L1737
I've not tried modifying this yet.

**Reply by codev123** [reply]
Sir do you have any code snippet where you have tried scanning channel before transmitting, with mentioned scheme applied. It will be big help. Thankyou

**Reply by GUVWAF** [reply]
Yes, it's here: https://github.com/meshtastic/firmware/blob/master/src/mesh/SX126xInterface.cpp#L289
